crates/ruff_linter/src/rules/ruff/rules/sort_dunder_all.rs

```diff
@@ -0,0 +1,883 @@
+        return;
+    }
+
+    let (elts, range, kind) = match node {
+        ast::Expr::List(ast::ExprList { elts, range, .. }) => (elts, *range, DunderAllKind::List),
+        ast::Expr::Tuple(tuple_node @ ast::ExprTuple { elts, range, .. }) => {
+            (elts, *range, DunderAllKind::Tuple(tuple_node))
+        }
+        _ => return,
+    };
+
+    let mut possibly_fixable = true;
+    let mut string_items = vec![];
+    for elt in elts {
+        // Don't flag `__all__` definitions that contain non-strings
+        let Some(string_literal) = elt.as_string_literal_expr() else {
+            return;
+        };
+        // If any strings are implicitly concatenated, don't bother trying to autofix
+        if string_literal.value.is_implicit_concatenated() {
+            possibly_fixable = false;
+        }
+        string_items.push(string_literal.value.to_str());
+    }
+    if dunder_all_is_already_sorted(&string_items) {
+        return;
+    }
+
+    let mut diagnostic = Diagnostic::new(UnsortedDunderAll, range);
+
+    if possibly_fixable {
+        if let Some(fix) = get_fix(range, elts, &string_items, &kind, checker) {
+            diagnostic.set_fix(fix);
+        }
+    }
+
+    checker.diagnostics.push(diagnostic);
+}
+
+enum DunderAllKind<'a> {
+    List,
+    Tuple(&'a ast::ExprTuple),
+}
+
+impl DunderAllKind<'_> {
+    fn opening_token_for_multiline_definition(&self) -> Tok {
+        match self {
+            Self::List => Tok::Lsqb,
+            Self::Tuple(_) => Tok::Lpar,
+        }
+    }
+
+    fn closing_token_for_multiline_definition(&self) -> Tok {
+        match self {
+            Self::List => Tok::Rsqb,
+            Self::Tuple(_) => Tok::Rpar,
+        }
+    }
+}
+
+fn dunder_all_is_already_sorted(string_elements: &[&str]) -> bool {
+    let mut element_iter = string_elements.iter();
+    let Some(this) = element_iter.next() else {
+        return true;
+    };
+    let mut this_key = AllItemSortKey::from(*this);
+    for next in element_iter {
+        let next_key = AllItemSortKey::from(*next);
+        if next_key < this_key {
+            return false;
+        }
+        this_key = next_key;
+    }
+    true
+}
+
+struct AllItemSortKey {
+    category: InferredMemberType,
+    value: String,
+}
+
+impl Ord for AllItemSortKey {
+    fn cmp(&self, other: &Self) -> Ordering {
+        self.category
+            .cmp(&other.category)
+            .then_with(|| natord::compare(&self.value, &other.value))
+    }
+}
+
+impl PartialOrd for AllItemSortKey {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+impl PartialEq for AllItemSortKey {
+    fn eq(&self, other: &Self) -> bool {
+        self.cmp(other) == Ordering::Equal
+    }
+}
+
+impl Eq for AllItemSortKey {}
+
+impl From<&str> for AllItemSortKey {
+    fn from(value: &str) -> Self {
+        Self {
+            category: InferredMemberType::of(value),
+            value: String::from(value),
```

Comment:
This obviously isn't ideal :(

I wanted to make this struct generic over a lifetime, i.e.

```rs
struct AllItemSortKey<'a> {
    category: InferredMemberType,
    value: &'a str,
}
```

But had trouble making it work.